libs/common-types/src/lib.rs

```diff
@@ -0,0 +1,270 @@
+		}
+	}
+
+	fn add(&mut self, property: Self::Property) {
+		match property {
+			PoolRole::Borrower => self.admin.insert(AdminRoles::BORROWER),
+			PoolRole::LiquidityAdmin => self.admin.insert(AdminRoles::LIQUIDITY_ADMIN),
+			PoolRole::PoolAdmin => self.admin.insert(AdminRoles::POOL_ADMIN),
+			PoolRole::PricingAdmin => self.admin.insert(AdminRoles::PRICING_ADMIN),
+			PoolRole::MemberListAdmin => self.admin.insert(AdminRoles::MEMBER_LIST_ADMIN),
+			PoolRole::RiskAdmin => self.admin.insert(AdminRoles::RISK_ADMIN),
+			PoolRole::TrancheInvestor(id, delta) => self.tranche_investor.insert(id, delta),
+		}
+	}
+}
+
+impl<Now, MaxTranches, MinDelay, TrancheId, Moment>
+	TrancheInvestors<Now, MaxTranches, MinDelay, TrancheId, Moment>
+where
+	MaxTranches: Get<TrancheId>,
+	Now: Time<Moment = Moment>,
+	MinDelay: Get<Moment>,
+	Moment: PartialEq + PartialOrd + Saturating + Ord + Copy,
+	TrancheId: PartialEq + PartialOrd,
+{
+	pub fn empty() -> Self {
+		Self::default()
+	}
+
+	pub fn is_empty(&self) -> bool {
+		self.info.is_empty()
+	}
+
+	fn validity(&self, delta: Moment) -> Moment {
+		let now = Now::now();
+		let min_validity = now.saturating_add(MinDelay::get());
+		let req_validity = now.saturating_add(delta);
+		max(min_validity, req_validity)
+	}
+
+	pub fn contains(&self, tranche: TrancheId) -> bool {
+		if tranche >= self.max_tranches {
+			return false;
+		}
+
+		self.info
+			.iter()
+			.position(|info| info.tranche_id == tranche && info.permissioned_till >= Now::now())
+			.is_some()
+	}
+
+	pub fn remove(&mut self, tranche: TrancheId, delta: Moment) {
+		if tranche >= self.max_tranches {
+			return;
+		}
+
+		if let Some(index) = self.info.iter().position(|info| info.tranche_id == tranche) {
+			let valid_till = &self.info[index].permissioned_till;
+			let now = Now::now();
+
+			if *valid_till <= now {
+				// The account is already invalid. Hence no more grace period
+				return;
+			} else {
+				// Ensure that permissioned_till is at least now + min_delay.
+				self.info[index].permissioned_till = self.validity(delta);
+			}
+		};
+	}
+
+	pub fn insert(&mut self, tranche: TrancheId, delta: Moment) {
+		if tranche >= self.max_tranches {
+			return;
+		}
+
+		let validity = self.validity(delta);
+
+		if let Some(index) = self.info.iter().position(|info| info.tranche_id == tranche) {
+			self.info[index].permissioned_till = validity;
+		} else {
+			self.info.push(TrancheInvestorInfo {
+				tranche_id: tranche,
+				permissioned_till: validity,
+			})
+		}
+	}
+}
+
+/// A struct we need as the pallets implementing trait Time
+/// do not implement TypeInfo. This wraps this and implements everything manually.
+#[derive(Encode, Decode, Eq, PartialEq, Debug, Clone)]
+pub struct TimeProvider<T, M>(PhantomData<(T, M)>);
+
+impl<T, Moment> Time for TimeProvider<T, Moment>
+where
+	T: Time<Moment = Moment>,
+	Moment: AtLeast32Bit + Parameter + Default + Copy,
+{
+	type Moment = Moment;
+
+	fn now() -> Self::Moment {
+		<T as Time>::now()
```

Comment:
This should be UnixTime instead of Time for getting seconds easily

Reply:
Ah okay. Will Change that. Do we default to using UnixTime?

Reply:
yeah since we only deal with seconds everywhere to be consistent across